aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStepTest.java

```diff
@@ -0,0 +1,239 @@
+        boolean found = false;
+        for (FlowNode n = walker.next(); n != null; n = walker.next()) {
+            if (n instanceof StepNode && ((StepNode) n).getDescriptor() instanceof SynchronousNonBlockingStep.DescriptorImpl) {
+                found = true;
+                break;
+            }
+        }
+
+        System.out.println("Checking flow node added...");
+        assertTrue("FlowNode has to be added just when the step starts running", found);
+
+        // Check for message the test message sent to context listener
+        System.out.println("Checking build log message present...");
+        j.waitForMessage("Test Sync Message", b);
+        // The last step did not run yet
+        j.assertLogContains("First message", b);
+        j.assertLogNotContains("Second message", b);
+
+        // Let syncnonblocking to continue
+        SynchronousNonBlockingStep.notify("wait");
+
+        System.out.println("Waiting until syncnonblocking (and the full flow) finishes");
+        j.waitForCompletion(b);
+        System.out.println("Build finished. Continue.");
+        // Check for the last message
+        j.assertLogContains("Second message", b);
+        j.assertBuildStatusSuccess(b);
+    }
+
+    @Test
+    public void interruptedTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "try { syncnonblocking 'wait' } catch(InterruptedException e) { echo 'Interrupted!' }\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        // Wait for syncnonblocking to be started
+        System.out.println("Waiting to syncnonblocking to start...");
+        SynchronousNonBlockingStep.waitForStart("wait", b);
+
+        // At this point syncnonblocking is waiting for an interruption
+
+        FlowExecution e = b.getExecutionPromise().get();
+        // Let's force a call to stop. This will try to send an interruption to the run Thread
+        e.interrupt(Result.ABORTED);
+        System.out.println("Looking for interruption received log message");
+        j.waitForMessage("Interrupted!", b);
+        System.out.println("Waiting for build completion");
+        j.waitForCompletion(b);
+    }
+
+    @Test
+    public void parallelTest() throws Exception {
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node {\n" +
+            "echo 'First message'\n" +
+            "parallel( a: { syncnonblocking 'wait0'; echo 'a branch'; }, b: { semaphore 'wait1'; echo 'b branch'; } )\n" +
+            "echo 'Second message'\n" +
+        "}"));
+        WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+
+        SynchronousNonBlockingStep.waitForStart("wait0", b);
+        SemaphoreStep.waitForStart("wait1/1", b);
```

Comment:
`success` may be called before the step is encountered; `SemaphoreStep` will just save the return value for later. Then we will `waitForMessage` anyway, proving that we got output from the second branch while the first branch was inside the synchronous step.

Nothing really wrong with the test as it is, just noting that there is a reason no other test calls `waitForStart` followed by `success` on the same ID with no intervening work.